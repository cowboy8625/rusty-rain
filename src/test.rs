use super::{Parser, Rain, cli::Cli};
use std::fmt::Write;

struct SnapshotOptions {
    label: String,
    cycles: usize,
    width: usize,
    height: usize,
    direction: super::Direction,
    chars: super::Characters,
}

impl Default for SnapshotOptions {
    fn default() -> Self {
        Self {
            label: String::new(),
            cycles: 25,
            width: 40,
            height: 20,
            direction: super::Direction::Down,
            chars: super::Characters::Bin,
        }
    }
}

fn display<const N: usize>(id: usize, window: &mut String, rain: &Rain<N>) {
    let width = rain.width;
    let height = rain.height;
    let char_width = rain.group.width() as usize;
    let id_str = format!("{:02X}", id); // hex with leading zero if needed
    write!(
        window,
        "{:-^width$}\n",
        id_str,
        width = width * char_width + 5
    )
    .unwrap();
    for (i, chunk) in rain.screen_buffer.chunks(width).enumerate() {
        write!(window, "{:02X} |", i).unwrap();
        write!(
            window,
            "{}|",
            &chunk
                .iter()
                .map(|c| c.display(char_width))
                .collect::<String>()
        )
        .unwrap();
        if i == height {
            continue;
        }
        write!(window, "\n").unwrap();
    }
}

fn set_up_snapshot(options: SnapshotOptions) {
    let SnapshotOptions {
        label,
        cycles,
        width,
        height,
        chars,
        direction,
    } = options;
    let mut cli = Cli::parse();
    cli.chars = chars;
    cli.direction = direction;
    let mut rain = Rain::<1024>::new(width, height, &cli);
    let mut window = String::new();
    for id in 0..cycles {
        rain.update();
        rain.update_screen_buffer().unwrap();
        display(id, &mut window, &rain);

        std::thread::sleep(std::time::Duration::from_millis(100));
    }
    insta::assert_snapshot!(label, window);
}

#[test]
fn test_screen_buffer() {
    set_up_snapshot(SnapshotOptions {
        label: "screen_buffer".to_string(),
        ..Default::default()
    });
}

#[test]
fn test_screen_buffer_direction_right_emoji_moon_double_width() {
    set_up_snapshot(SnapshotOptions {
        label: "screen_buffer_direction_right_emoji_moon_double_width".to_string(),
        cycles: 50,
        width: 32,
        height: 10,
        direction: super::Direction::Right,
        chars: super::Characters::Moon,
        ..Default::default()
    });
}

#[test]
fn test_screen_buffer_direction_left_emoji_crab_double_width() {
    set_up_snapshot(SnapshotOptions {
        label: "test_screen_buffer_direction_left_emoji_crab_double_width".to_string(),
        cycles: 50,
        width: 32,
        height: 10,
        direction: super::Direction::Left,
        chars: super::Characters::Crab,
        ..Default::default()
    });
}

#[test]
fn test_gen_shade_color() {
    use super::{Color, gen_shade_color};
    use pretty_assertions::assert_eq;
    let bc = Color::Rgb { r: 0, g: 255, b: 0 };
    let length = 10;

    let colors = gen_shade_color(bc, length);

    assert_eq!(colors.len(), length as usize);
    assert_eq!(colors.first(), Some(&Color::Rgb { r: 0, g: 225, b: 0 }));
    assert_eq!(colors.last(), Some(&Color::Rgb { r: 0, g: 0, b: 0 }));
}
