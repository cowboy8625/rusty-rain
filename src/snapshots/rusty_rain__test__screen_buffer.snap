---
source: src/test.rs
expression: window
---
---------------------00----------------------
00 |                                        |
01 |                                        |
02 |                                        |
03 |                                        |
04 |                                        |
05 |                                        |
06 |                                        |
07 |                                        |
08 |                                        |
09 |                                        |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------01----------------------
00 |                                        |
01 |                                        |
02 |                                        |
03 |                                        |
04 |                                        |
05 |                                        |
06 |                                        |
07 |                                        |
08 |                                        |
09 |                                        |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------02----------------------
00 |              11  0              0      |
01 |                                        |
02 |                                        |
03 |                                        |
04 |                                        |
05 |                                        |
06 |                                        |
07 |                                        |
08 |                                        |
09 |                                        |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------03----------------------
00 |              11  0              0      |
01 |              01                 1      |
02 |                                        |
03 |                                        |
04 |                                        |
05 |                                        |
06 |                                        |
07 |                                        |
08 |                                        |
09 |                                        |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------04----------------------
00 |  1           11  0  1           0      |
01 |              01  1              1      |
02 |              10                 1      |
03 |                                        |
04 |                                        |
05 |                                        |
06 |                                        |
07 |                                        |
08 |                                        |
09 |                                        |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------05----------------------
00 | 01         0 11  0  1           0      |
01 |              01  1  1           1      |
02 |              10                 1      |
03 |              00                 1      |
04 |                                        |
05 |                                        |
06 |                                        |
07 |                                        |
08 |                                        |
09 |                                        |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------06----------------------
00 | 01         0011  0  1           0 1    |
01 |  1           01  1  1           1      |
02 |              10  0  0           1      |
03 |              00                 1      |
04 |              11                 1      |
05 |                                        |
06 |                                        |
07 |                                        |
08 |                                        |
09 |                                        |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------07----------------------
00 | 01         0011  0  1           0 1    |
01 | 01         0001  1  1           1      |
02 |              10  0  0           1      |
03 |              00     1           1      |
04 |              11                 1      |
05 |              01                 1      |
06 |                                        |
07 |                                        |
08 |                                        |
09 |                                        |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------08----------------------
00 | 01         0011  0  1     1   1   1    |
01 | 01         0001  1  1           1 0    |
02 |  1          010  0  0           1      |
03 |              00  0  1           1      |
04 |              11     1           1      |
05 |              01                 1      |
06 |              10                 0      |
07 |                                        |
08 |                                        |
09 |                                        |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------09----------------------
00 | 01         0011  0 01     1   1   1    |
01 | 01         0001  1  1     0   1   0    |
02 | 01         1010  0  0           1      |
03 |             100  0  1           1      |
04 |              11     1           1      |
05 |              01     1           1      |
06 |              10                 0      |
07 |              10                 0      |
08 |                                        |
09 |                                        |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------0A----------------------
00 | 01    1    0011  0 01     1   1   1 0  |
01 | 01         0001  1  1     0   1   0    |
02 | 01         1010  0  0     1   1   1    |
03 |  1          100  0  1           1      |
04 |             011  0  1           1      |
05 |              01     1           1      |
06 |              10     1           0      |
07 |              10                 0      |
08 |              01                 0      |
09 |                                        |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------0B----------------------
00 | 01 0  1 0  0011  0 01     1   1   1 01 |
01 | 01         0001  1 11     0   1   0 0  |
02 | 01         1010  0  0     1   1   1    |
03 | 11         1100  0  1     1   0        |
04 |             011  0  1           1      |
05 |             101     1           1      |
06 |              10     1           0      |
07 |              10     1           0      |
08 |              01                 0      |
09 |              01                 1      |
0A |                                        |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------0C----------------------
00 | 01 0  1 0 1001   0 0      1       1001 |
01 | 01    1    0001  1 11     0   1   0 00 |
02 | 01         1010  0  0     1   1   1 1  |
03 | 11         1100  0  1     1   0   0    |
04 |  0          011  0  1     0   1        |
05 |             101  0  1           1      |
06 |             010     1           0      |
07 |              10     1           0      |
08 |              01     1           0      |
09 |              01                 1      |
0A |              11                 0      |
0B |                                        |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------0D----------------------
00 | 01 0 11 0 100    0 0      1       1001 |
01 | 01 0  1 0  000   1 1      0       0 00 |
02 | 01         1010  0 00     1   1   1 10 |
03 | 11         1100  0  1     1   0   0 0  |
04 | 00         1011  0  1     0   1        |
05 |             101  0  1     1   0        |
06 |             010     1           0      |
07 |             010     1           0      |
08 |              01     1           0      |
09 |              01     1           1      |
0A |              11                 0      |
0B |              10                 1      |
0C |                                        |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------0E----------------------
00 |101 0 11 0 100      0  0   1       1001 |
01 | 01 0  1 0 100    1 1      0       0100 |
02 | 01    0    101   0 0      1       1 10 |
03 | 11         1100  0  1     1   0   0 01 |
04 | 00         1011  0  1     0   1   1 0  |
05 |  0          101  0  1     1   0        |
06 |             010  0  1     1   0        |
07 |             010     1           0      |
08 |             101     1           0      |
09 |              01     1           1      |
0A |              11     1           0      |
0B |              10                 1      |
0C |              11                 1      |
0D |                                        |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------0F----------------------
00 |101 0 1110 100      0  0  1       01001 |
01 | 01 0 01 0 100    1 1      0       0100 |
02 | 01 1  0 1  10    0 0      1       1 10 |
03 | 11         110   0 1      1       0 01 |
04 | 00         1011  0  1     0   1   1 00 |
05 | 10         1101  0  1     1   0     0  |
06 |             010  0  1     1   0        |
07 |             010     1     1   1        |
08 |             101     1           0      |
09 |             001     1           1      |
0A |              11     1           0      |
0B |              10     0           1      |
0C |              11                 1      |
0D |              00                 1      |
0E |                                        |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------10----------------------
00 |101 0 1110 100      0 00  1       01001 |
01 |101 0 0110 100      1  0  0        0100 |
02 | 01 1  0 1 010    0 0      1       1010 |
03 | 11    0    11    0 1      1       0 01 |
04 | 00         101   0        0       1 00 |
05 | 10         1101  0  1     1   0   0 00 |
06 |  0          010  0  1     1   0     1  |
07 |             010  0  1     1   1        |
08 |             101     1     0   0        |
09 |             001     1           1      |
0A |             111     1           0      |
0B |              10     0           1      |
0C |              11     0           1      |
0D |              00                 1      |
0E |              11                 1      |
0F |                                        |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------11----------------------
00 |101 001110 100      0 00  1       010010|
01 |101 0 0110 100      1  0  0       10100 |
02 | 01 1 1001 010    0 0     0        1010 |
03 | 11 0  0 0  11    0 1      1       0 01 |
04 | 00         10    0 1      0       1 00 |
05 | 10         110   0        1       0 00 |
06 | 10         0010  0  1     1   0     10 |
07 |             010  0  1     1   1     1  |
08 |             101     1     0   0        |
09 |             001     1     0   0        |
0A |             111     1           0      |
0B |             110     0           1      |
0C |              11     0           1      |
0D |              00     0           1      |
0E |              11                 1      |
0F |              01                 0      |
10 |                                        |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------12----------------------
00 |1010001110 100   0  0 00 11  1    010010|
01 |101 0 0110 100      1 10  0       101000|
02 |001 1 1001 010      0  0  0        1010 |
03 | 11 0  010 011    0 1     0        0001 |
04 | 00    0    10    0 1      0       1 00 |
05 | 10         11    0        1       0 00 |
06 | 10         001   0        1       0 10 |
07 |  1          010  0  1     1   1     11 |
08 |             101  1  1     0   0     1  |
09 |             001     1     0   0        |
0A |             111     1     0   1        |
0B |             110     0           1      |
0C |             111     0           1      |
0D |              00     0           1      |
0E |              11     1           1      |
0F |              01                 0      |
10 |              01                 0      |
11 |                                        |
12 |                                        |
13 |                                        |
---------------------13----------------------
00 |10100011 0 10    0  0 00 11 01    010010|
01 |1011000110 100   1  1 10  0       101000|
02 |001 1 1001 010      0  0  0       110100|
03 | 11 0 0010 011    0 1     0        0001 |
04 | 00 1  010  10    0 1     1        1 00 |
05 | 10         11    0 1      1       0 00 |
06 | 10         00    0        1       0 10 |
07 | 01         001   0        1         11 |
08 |             101  1  1     0   0     11 |
09 |             001     1     0   0     1  |
0A |             111     1     0   1        |
0B |             110     0     1   1        |
0C |             111     0           1      |
0D |             100     0           1      |
0E |              11     1           1      |
0F |              01     1           0      |
10 |              01                 0      |
11 |              11                 1      |
12 |                                        |
13 |                                        |
---------------------14----------------------
00 |10100011 0010   10 10 00 11 01  1 010010|
01 |10110001 0 10    1  1 10 10 00    101000|
02 |00111 1001 010   1  0 00  0       110100|
03 |011 0 0010 011      1  1  0        00010|
04 | 00 1  010 110    0 1     1        1000 |
05 | 10    11   11    0 1     0        0 00 |
06 | 10         00    0        1       0 10 |
07 | 01         00    0        1       0 11 |
08 |  0          10   1        0         11 |
09 |             001  0  1     0   0     11 |
0A |             111     1     0   1     1  |
0B |             110     0     1   1        |
0C |             111     0     1   0        |
0D |             100     0           1      |
0E |             011     1           1      |
0F |              01     1           0      |
10 |              01     1           0      |
11 |              11                 1      |
12 |              11                 1      |
13 |                                        |
---------------------15----------------------
00 |10100011 0010   10 10 00 11 01  1 010010|
01 |10110001 0 10    1 01 10 10 00    101000|
02 |00111010 1 01    1  0 00  0 1     110100|
03 |01110 0010 011   0  1  1  0       000010|
04 | 00 1 0010 110    0 1     1        10001|
05 | 10 1  110  11    0 1     0        0 00 |
06 | 10     1   00    0 0     0        0 10 |
07 | 01         00    0        1       0 11 |
08 | 10         01    1        0         11 |
09 |             00   0        0         11 |
0A |             111     1     0   1     10 |
0B |             110     0     1   1     0  |
0C |             111     0     1   0        |
0D |             100     0     1   1        |
0E |             011     1           1      |
0F |             101     1           0      |
10 |              01     1           0      |
11 |              11     0           1      |
12 |              11                 1      |
13 |              01                 1      |
---------------------16----------------------
00 |10100011 0010   1  10 00011 010 1 010010|
01 |10110001 0010   01 01 10 10 00  0 101000|
02 |00111010 1 01    1 00 00 00 11    110100|
03 |01110 00 0 01    0  1 01  0 1     000010|
04 |00011 0010 110   1  1  1  1        10001|
05 | 10 1  110 011    0 1     0        00000|
06 | 10    01   00    0 0     0        0 10 |
07 | 01     0   00    0       0        0 11 |
08 | 10         01    1        0       0 11 |
09 |  1          0    0        0         11 |
0A |             11   1        0         10 |
0B |             110     0     1   1     01 |
0C |             111     0     1   0     0  |
0D |             100     0     1   1        |
0E |             011     1     1   0        |
0F |             101     1           0      |
10 |             101     1           0      |
11 |              11     0           1      |
12 |              11     0           1      |
13 |              01                 1      |
---------------------17----------------------
00 |10100011 0010   1  10 00011 010 1 010 10|
01 |10110001 0010   0  01 10 10 00  0 101000|
02 |00111010 1 01    1 00 00 00 11    110100|
03 |01110000 0 01    0 01 01  0 1     000010|
04 |00011 00 0 11    1  1  1  1 1     010001|
05 | 1001 1110 011   10 1     0        00000|
06 | 10 0  011  00    0 0     0        0 101|
07 | 01     0   00    0 1     0        0 11 |
08 | 10     0   01    1       1        0 11 |
09 | 11         10    0        0         11 |
0A |             1    1        0         10 |
0B |             11            1         01 |
0C |             111     0     1   0     00 |
0D |             100     0     1   1     1  |
0E |             011     1     1   0        |
0F |             101     1     1   0        |
10 |             101     1           0      |
11 |             011     0           1      |
12 |              11     0           1      |
13 |              01     1           1      |
---------------------18----------------------
00 |10100011 0010   1  10 00011 010 1 010 1 |
01 |10110001 0010   0  01 10010 001 0 101 00|
02 |00111010 1001   0  00 00 00 11  0 110100|
03 |01110000 0 01    0 01 01 00 10    000010|
04 |00011 00 0 11    1 11 01  1 1     010001|
05 |11001 11 0 01    1  1  0  0 1      00000|
06 | 1000  011 100   10 0     0        00101|
07 | 01    00   00    0 1     0        0 111|
08 | 10     0   01    1       1        0 11 |
09 | 11     1   10    0       1        1 11 |
0A |  1          1    1        0         10 |
0B |             1    1        1         01 |
0C |             11            1         00 |
0D |             100     0     1   1     11 |
0E |             011     1     1   0     0  |
0F |             101     1     1   0        |
10 |             101     1     0   1        |
11 |             011     0           1      |
12 |             011     0           1      |
13 |              01     1           1      |
